src/components/document/editable-document-content.tsx

```diff
@@ -114,6 +115,33 @@ export function EditableDocumentContent({
     contained ? "contained-editable-document-content" : "full-screen-editable-document-content",
     {"comment-select" : documentSelectedForComment, "full-height": fullHeight}, className);
 
+  useEffect(() => {
+    const disposer = onSnapshot(document, (snapshot: DocumentModelSnapshotType) => {
+      const tileMap = snapshot.content?.tileMap || {};
+      const tileTypes: string[] = [];
+
+      Object.keys(tileMap).forEach((key) => {
+        const tileInfo = tileMap[key] as ITileMapEntry;
+        const type = tileInfo.content.type;
+        if (!tileTypes.includes(type)) {
+          tileTypes.push(type);
+        }
+      });
```

Comment:
I'd suggest moving this logic into `DocumentContentModel` as a `get tileTypeList` view.  That way it should get cached and only recalculated when things change.

Reply:
Will this work even though it's technically getting the tiles from the snapshot, not the document?

Reply:
@bgoldowsky This has moved into use-document-sync-to-firebase, but the question still remains. We could work with the `document` object that is being watched for snapshots. However I'm not sure how safe that is. The snapshot calls might be async so the `document` object might not match the snapshot when the onSnapshot callback is called. So I'd vote to keep this working with the snapshot. It could still be extracted to a function though, and probably that function could probably work either with a snapshot or the model itself.

Reply:
In #2368 I made a function in `base-document-content.ts`. But if this is operating on a snapshot that can't be used here.  It seems ok to leave them separate.